File: SICP/Lecture_2/Code/Java/CompoundData/src/fayelab/sicp/compounddata/rationalnumber/RationalNumberWithoutGcd.java
package fayelab.sicp.compounddata.rationalnumber;

import static java.util.Arrays.asList;

import java.util.List;

public class RationalNumberWithoutGcd
{
    public static List<Integer> makeRationalNumber(int numer, int denom)
    {
        return asList(numer, denom);
    }
    
    public static int numer(List<Integer> rationalNumber)
    {
        return head(rationalNumber);
    }

    public static int denom(List<Integer> rationalNumber)
    {
        return tail(rationalNumber);
    }
    
    public static List<Integer> add(List<Integer> rationalNumber1, List<Integer> rationalNumber2)
    {
        return makeRationalNumber(numer(rationalNumber1) * denom(rationalNumber2) + denom(rationalNumber1) * numer(rationalNumber2),
                                  denom(rationalNumber1) * denom(rationalNumber2));
    }
    
    public static List<Integer> mul(List<Integer> rationalNumber1, List<Integer> rationalNumber2)
    {
        return makeRationalNumber(numer(rationalNumber1) * numer(rationalNumber2),
                                  denom(rationalNumber1) * denom(rationalNumber2));
    } 

    private static int head(List<Integer> pair)
    {
        return pair.get(0);
    }
    
    private static int tail(List<Integer> pair)
    {
        return pair.get(1);
    }
}

File: SICP/Lecture_2/Code/Python/compound_data/rational_number/rational_number_without_gcd.py
def make_rational_number(_numer, _denom):
    return [_numer, _denom]

def numer(rational_number):
    return head(rational_number)

def denom(rational_number):
    return tail(rational_number)

def add(rational_number1, rational_number2):
    return make_rational_number(
        numer(rational_number1) * denom(rational_number2) + denom(rational_number1) * numer(rational_number2),
        denom(rational_number1) * denom(rational_number2))

def mul(rational_number1, rational_number2):
    return make_rational_number(
        numer(rational_number1) * numer(rational_number2),
        denom(rational_number1) * denom(rational_number2))

def head(pair):
    return pair[0]

def tail(pair):
    return pair[1]